File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-3.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        window_start = 0
        window_end = len(nums) - 1

        if len(nums) == 1:
            if target != nums[0]:
                return -1
            else:
                return 0

        while window_start <= window_end:
            middle = (window_start + window_end) // 2

            if nums[middle] == target:
                return middle

            if nums[window_start] <= nums[middle]:
                if nums[window_start] <= target < nums[middle]:
                    window_end = middle - 1
                else:
                    window_start = middle + 1
            else:
                if nums[middle] < target <= nums[window_end]:
                    window_start = middle + 1
                else:
                    window_end = middle - 1



        
        return -1

File: Data Structures & Algorithms/two-integer-sum/submission-3.java
// class Solution:
//     def twoSum(self, nums: List[int], target: int) -> List[int]:
//         # Store the value, and the other value needed to be valid
//         val_map = {}
//         for i in range(len(nums)):
//             # Try and see if this value is a target
//             if nums[i] in val_map.keys():
//                 return [val_map[nums[i]], i]
//             # Add val and target
//             val_map[target - nums[i]] = i


            


class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> val_map = new HashMap<>();

        for (int i = 0; i < nums.length; i++)
        {
            if (val_map.keySet().contains(nums[i]))
            {
                return new int[]{val_map.get(nums[i]), i};
            }

            val_map.put(target - nums[i], i);
        }

        return new int[]{};
    }
}
